# Extensible Exceptions

In this post, I'm going to use the term "exception" to refer to returning
a result that indicates that something went wrong. I'm not talking about
the throw/catch machinery that many high-level language offer. I'm talking
about exceptions that would be returned like any other value. For example:

    read : FileDescriptor -> Int -> IO (Either Errno Bytes)

I read a [discussion of exceptions](https://www.reddit.com/r/haskell/comments/hzrmpz/forcing_exceptions_to_be_listed_in_the_type/)
on reddit the other day that got me thinking about this. There is disagreement
in that post about whether throw/catch is good or bad. Personally I'm a
proponent of having uncatchable `panic` that kills all threads in the
program with no cleanup, and I think anything that doesn't fall in the
"unrecoverable" category should just be returned as a value (as in `read`
above). What I'm curious about is whether it is possible to improve the
ergonomics of this.

In Haskell, with lower-level code, one particular problem I've found is that
it can be tricky to deal with error codes from the operating system. Why?
System calls often result a C `int` through `errno`. All the possible `errno`
values are [documented](https://www-numi.fnal.gov/offline_software/srt_public_context/WebDocs/Errors/unix_system_errors.html),
but not every call can actually return every `errno` value. The Linux
man pages [list several errors that `read` can return](https://man7.org/linux/man-pages/man2/read.2.html).
Among these are: `EAGAIN`, `EBADF`, `EINVAL`, `EIO`. Notably absent is
`ECONNRESET` (possible when reading from a TCP socket), but there's a
catch-all in `read`'s documentation:

    Other errors may occur, depending on the object connected to fd.

Alright. So, perhaps anything is fair game. Not quite. Errors like `EMLINK`
and `EXDEV` are certainly not going to happen because those only happen
when creating links on the file system. So really what's going on is that
there is some kind of nebulous, context-dependent set of errors that can
happen any time we make a system call. To understand which errors can happen,
you have to understand both the system call the type of file descriptor that
it is working on. For example:

    int fd = open("/path/to/file", O_RDONLY);
    ... remember to handle open errors
    fstat(fd,&statbuffer)
    ... handle fstat errors
    if(S_ISREG(statbuffer.st_mode)) {
      ssize_t n = read(fd, &buffer, 4096))
    } else { ... fail on non-regular files ... }

In this context, we know that the call to `read` cannot return `ECONNRESET`.
We know this because we made sure it was a regular file, not a domain socket.
But what if it somehow return `ECONNRESET`? What's the right thing to do?
Panic. It's unrecoverable. One of two things has happened:

1. The operating system has violated its contract with the programmer.
2. The programmer misunderstands the operating system. (more likely)

In either case, there is no way to handle the error. It is senseless to just
wait for some number of milliseconds and try again becaues the programmer
themself does not understand what has happened. The program needs to report
the error with as much detail as possible and die and then the programmer
can fix the application.

What about something like `EBADF`? This also should never happen. If it does,
it's the programmer's fault. `EBADF` should be handled with a panic since
it indicates that the application has mismanaged file descriptors.

There are also error codes like `ENOMEM`, which are somewhere in the middle
concerning recoverability. `ENOMEM` roughly means that the operating system
is out of memory. I've never actually recovered from an `ENOMEM` error, and
it would be even more difficult in a high-level language than in C. I suspect
that you could try to release a bunch of unneeded resources (and maybe
run a compacting GC and `munmap` some memory to hand it back to the OS).
It seems unlikely that an application written in a high-level language
would go to the trouble to recover from this, but it is possible, and
the application is not necessarily broken if this happens. So I think of
it as a grey area.

So let's get back to higher-level languages. In C, it is not possible to
use the type system to refine knowledge about an error. Even in most
higher-level languages, this is not trivial and certainly not ergonomic.
Here is a possibility that uses something like a refinement type:

    type Errno = Int
    -- The ReadErrno type includes anything that could happen that the
    -- could not indicate an broken application.
    type ReadErrno = {EAGAIN, EIO, ECONNRESET, ... }
    c_read : FileDescriptor -> Int -> IO (Either Errno Bytes)
    read : FileDescriptor -> Int -> IO (Either ReadErrno Bytes)
    read fd n = do
      r <- c_read fd n
      case r of
        Left err -> case downcast @Errno @ReadErrno err of
          Left err' -> panic (describe err')
          Right err' -> pure (Left err')
        Right b -> pure (Right b)

Above, `read` refines `c_read` by panicing on any errors that could not
by handled by a user. It may be the case that the end user of `read`
wants to panic on even more errors. For example, if they believe they
will be working on a regular file, they may want:

    type ReadRegularErrno = {EAGAIN, EIO, ... }
    readRegular : FileDescriptor -> Int -> IO (Either ReadRegularErrno Bytes)
    readRegular fd n = do
      r <- read fd n
      case r of
        Left err -> case downcast @ReadErrno @ReadRegularErrno err of
          Left err' -> panic (describe err')
          Right err' -> pure (Left err')
        Right b -> pure (Right b)

This feels pretty repetitive. There is probably a way to eliminate some
of this repetition, but I'm not going to worry about that for now. It
is possible to eliminate error codes without just panicing. Commonly,
nonblocking IO will be integrated with an event manager:

    type ReadCompleteErrno = ReadErrno - {EAGAIN}
    readComplete : FileDescriptor -> Int -> IO (Either ReadCompleteErrno Bytes)
    readComplete fd n = do
      r <- read fd n
      case r of
        Left err -> case err of   -- magic refining case
          EAGAIN -> do
            waitUntilReady fd
            readComplete fd n
          err' -> pure (Left err')
        Right b -> pure (Right b)

Pretty neat. While I believe a system like this could help when interface
with the operating system, I'm less sure of its value more broadly. For
this to have value, you need to be in a situation where error handling
requiring plucking out certain types of errors and leaving others for
something at a higher level to handle.

Consider [HttpExceptionContent](https://hackage.haskell.org/package/http-client-0.7.1/docs/Network-HTTP-Client.html#t:HttpExceptionContent)
from the `http-client` library. This lists a bunch of things that can go
wrong when issuing an HTTP request. A user might want to handle some of
these differently from others, but it's unlikely that a user would want
to handle just one or two and then let the function it returns to handle
the rest. All the handling (including failing hard or just logging what
happened) is likely to happen in the same place, so extensible exceptions
do not really buy anything here. A plain old sum does just fine.

It could be argued that a user of `http-client` may perform other actions
over the network after issueing an HTTP request and that it would be to
their benefit to aggregate all possible errors into a big sum. I disagree
with this argument. By doing this kind of `ExceptT`-style handling, the
programmer loses access to useful context about the error messages.
Better would be to include this context:

    data MyError
      = ShoppingCartApiError HttpException
      | TwitterApiError HttpException
      | DatabaseError DatabaseException
    action : ExceptT MyError IO Foo
    action = do
      withExceptT ShoppingCartApiError $ ...
      withExceptT TwitterApiError $ ...
      withExceptT DatabaseError $ ...

But even better would be to just log the exceptions here. Unless you are
trying to internationalize your error messages, you could just print them
out in `action`. The type would instead be:

    action : Logger -> IO (Either () Foo)
